llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp

```diff
@@ -3365,6 +3365,24 @@ SIRegisterInfo::getConstrainedRegClassForOperand(const MachineOperand &MO,
   return nullptr;
 }
 
+unsigned SIRegisterInfo::getNumSupportedRegs(const MachineFunction &MF) const {
+#ifndef NDEBUG
+  for (unsigned K = AMDGPU::AGPR0; K < AMDGPU::NUM_TARGET_REGS; ++K) {
+    // Skip lo16 registers, they're "fake" and don't have a regclass assigned.
+    if (K >= AMDGPU::AGPR0_HI16 && K <= AMDGPU::AGPR255_HI16)
+      continue;
+    if (!isAGPR(MF.getRegInfo(), K))
+      report_fatal_error("register at index " + Twine(K) + " is not an AGPR!");
+  }
+#endif
+
+  // Don't include AGPRs on targets that don't have them.
+  // This cuts about 4000 register (almost half of all registers) off.
+  return MF.getInfo<SIMachineFunctionInfo>()->mayUseAGPRs(MF.getFunction())
```

Comment:
Not sure how reliable this check is, maybe it should also check for a target feature ?

Reply:
It's not a target feature, it's a property of the specific function

Reply:
Actually this is bugged. You should probably just start with considering ST.hasGFX90AInsts(), but you want MFI::usesAGPRs, not mayUseAGPRs which looks at the underlying IR attribute.

I would also hope you wouldn't need to special case this. All the registers are reserved in the case where they are unavailable. 

The second possible bug is amdgpu-no-agprs means the IR doesn't require the kernel to allocate AGPRs. But AGPRs may still appear as a result of the occupancy bounds (we probably should fix the inference logic to consider that case)

Reply:
I haven't seen this getNumSupportedRegs callback before; it probably shouldn't exist? It seems redundant with RegisterClassInfo, which is supposed to filter out a lot of the redundant cases

Reply:
> Actually this is bugged. You should probably just start with considering ST.hasGFX90AInsts(), but you want MFI::usesAGPRs, not mayUseAGPRs which looks at the underlying IR attribute.

So hasGFX90AInsts && usesAGPR ?

> The second possible bug is amdgpu-no-agprs means the IR doesn't require the kernel to allocate AGPRs. But AGPRs may still appear as a result of the occupancy bounds (we probably should fix the inference logic to consider that case)

This is worrying because if we have to always enable AGPRs on all GFX90AInsts, then this patch is more or less useless. I think the large inputs that'd benefit from this often target MI and if MI targets don't benefit from it then I'd rather not bother

Would the fix be easy? Is there an extra check this can do to avoid the issue?

> I haven't seen this getNumSupportedRegs callback before; it probably shouldn't exist? It seems redundant with RegisterClassInfo, which is supposed to filter out a lot of the redundant cases

I think this is targeted at reducing iteration, and the size of containers that are used for registers (arrays, BitVectors, etc.). If this lands, I'd like to see if we can `getNumSupportedRegs` in more places.

Note that I felt like this was a good idea, but if you think it's too confusing and not worth it then I don't want to force it. I've been looking at ways to improve performance with backend passes for AMDGPU because some of them scale very poorly with a large amount of registers like we have, so reducing the amount of registers we're dealing with seemed like a good idea.

Reply:
> This is worrying because if we have to always enable AGPRs on all GFX90AInsts

90a insts was the wrong one, ST.hasMAIInsts && usesAGPRs. On gfx90a+ we hardly use AGPRs. the IR attribute only detects inline asm usage. We only use AGPRs there for very low occupancy cases (and I think requires manually overriding with amdgpu-waves-per-eu to ever use them)

> Would the fix be easy? Is there an extra check this can do to avoid the issue?

I don't think it matters for your purposes, I've just been thinking about the meaning of "amdgpu-no-agprs" and it's possible misinterpretation, which you seem to have run into here. 

> I think this is targeted at reducing iteration, and the size of containers that are used for registers (arrays, BitVectors, etc.). If this lands, I'd like to see if we can getNumSupportedRegs in more places.

It's too coarse grained. What we really want is to shrink the size of all classes to avoid reserved cases (which is the main point of RegisterClassInfo, which should probably be used more). This patch is probably fine as-is, but it's more limited than It could be (and could break if the register enum values ever change)

Reply:
Although actually usesAGPRs should just always be false for the no-AGPR targets, so it should be sufficient